zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs

```diff
@@ -0,0 +1,358 @@
+    },
+    table::CallContextFieldTag,
+    util::Expr,
+};
+use eth_types::{evm_types::OpcodeId, Field, ToScalar, U256};
+
+use halo2_proofs::{circuit::Value, plonk::Error};
+
+#[derive(Clone, Debug)]
+pub(crate) struct ErrorReturnDataOutOfBoundGadget<F> {
+    opcode: Cell<F>,
+    memory_offset: Cell<F>,
+    sum: AddWordsGadget<F, 2, true>,
+    /// Holds the size of the last callee return data.
+    return_data_length: Cell<F>,
+
+    is_data_offset_within_range: IsZeroGadget<F>,
+    // end = data_offset + length
+    is_end_within_range: IsZeroGadget<F>,
+    // when `end` not overflow, check if it exceeds return data size.
+    is_end_exceed_length: LtGadget<F, N_BYTES_U64>,
+    rw_counter_end_of_reversion: Cell<F>,
+    restore_context: RestoreContextGadget<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for ErrorReturnDataOutOfBoundGadget<F> {
+    const NAME: &'static str = "ErrorReturnDataOutOfBoundGadget";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::ErrorReturnDataOutOfBound;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+        let memory_offset = cb.query_cell();
+        let data_offset = cb.query_word_rlc();
+        let length = cb.query_word_rlc();
+        let end = cb.query_word_rlc();
+        let rw_counter_end_of_reversion = cb.query_cell();
+
+        let return_data_length = cb.query_cell();
+
+        cb.require_equal(
+            "opcode is RETURNDATACOPY",
+            opcode.expr(),
+            OpcodeId::RETURNDATACOPY.expr(),
+        );
+
+        // Pop memory_offset, offset, length from stack
+        cb.stack_pop(memory_offset.expr());
+        cb.stack_pop(data_offset.expr());
+        cb.stack_pop(length.expr());
+
+        // read last callee return data length
+        cb.call_context_lookup(
+            false.expr(),
+            None,
+            CallContextFieldTag::LastCalleeReturnDataLength,
+            return_data_length.expr(),
+        );
+
+        // check `data_offset` u64 overflow
+        let data_offset_larger_u64 = from_bytes::expr(&data_offset.cells[8..]);
+        let is_data_offset_within_range = IsZeroGadget::construct(cb, data_offset_larger_u64);
+
+        // when data set is not overflow, check sum and `end` if not overflow.
+        let (sum, is_end_within_range) = cb.condition(is_data_offset_within_range.expr(), |cb| {
+            let sum_gadget =
+                AddWordsGadget::construct(cb, [data_offset.clone(), length], end.clone());
+            let end_larger_u64 = from_bytes::expr(&end.cells[8..]);
+            (sum_gadget, IsZeroGadget::construct(cb, end_larger_u64))
+        });
+
+        // when `end` not u64 overflow, check if it exceeds return data length
+        let is_end_exceed_length = cb.condition(
+            is_data_offset_within_range.expr() * is_end_within_range.expr(),
+            |cb| LtGadget::construct(cb, return_data_length.expr(), end.expr()),
```

Comment:
We should use `from_bytes here, and I'd suggest to add a test case to have `return_data_length` larger than `0xff`.

```suggestion
            |cb| LtGadget::construct(cb, return_data_length.expr(), from_bytes::expr(&end.cells[..N_BYTES_U64])),
```

Reply:
will add test case for that.  be curious about why special care about `0xff`.

Reply:
`end.expr()` is random linear combination, but here we are comparing numbers. The test passes now just because the value is a byte so its RLC happens to be equal to its number repr.

Add some tests that are larger than a byte could avoid this kind of issue

Reply:
updated and add test in [f3bc7ac](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1017/commits/f3bc7ac6898a06c6ef724f12875b74410bb0a81d) thanks for catching.